### PR TITLE
Add missing jpa 3.2 features on correct branch

### PR DESCRIPTION
**Pull request by starksm64**
This replaces the https://github.com/eclipse-ee4j/glassfish/pull/25208 PR and uses the 8.0-JDK17 branch

### CODE REVIEW

File: appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/xml/PersistenceTagNames.java [hunk 1]
```diff
@@ -40,5 +40,7 @@ public interface PersistenceTagNames extends TagNames {
     public static String TRANSACTION_TYPE = "transaction-type";
     public static String SHARED_CACHE_MODE = "shared-cache-mode";
     public static String VALIDATION_MODE = "validation-mode";
+    public static String SCOOPE = "scope";
```
Comment (pzygielo) [inline appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/xml/PersistenceTagNames.java:43]: :eyes:

File: appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/node/PersistenceUnitNode.java [hunk 2]
```diff
@@ -113,6 +113,8 @@ private void initDispatchTable() {
         table.put(PersistenceTagNames.CLASS, "addClass");
         table.put(PersistenceTagNames.SHARED_CACHE_MODE, "setSharedCacheMode");
         table.put(PersistenceTagNames.VALIDATION_MODE, "setValidationMode");
+        table.put(PersistenceTagNames.SCOOPE, "setValidationMode");
```
Comment (pzygielo) [inline appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/node/PersistenceUnitNode.java:116]: :eyes: `setValidationMode`?

File: appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/annotation/handlers/EntityManagerReferenceQualifiedHandler.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
+ * Copyright (c) 1997, 2020 Oracle and/or its affiliates. All rights reserved.
```
Comment (pzygielo) [inline appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/annotation/handlers/EntityManagerReferenceQualifiedHandler.java:3]: Looks like new file, so (c) for 2024 only, and no Oracle would do, I guess.

File: appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/xml/PersistenceTagNames.java (hunk 1 above)
Comment (edburns) [reply appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/xml/PersistenceTagNames.java:43]: SCOOPE?

**Review by edburns — CHANGES_REQUESTED** [review]
Change `PersistenceTagNames.SCOOPE` to `PersistenceTagNames.SCOPE`.

**Comment by arjantijms** [comment]
> This replaces the https://github.com/eclipse-ee4j/glassfish/pull/25208 PR and uses the 8.0-JDK17 branch

It's really the 8.0 branch this should be in, although I can cross-merge.

8.0 is for 8.0 development. Just prior to releasing an 8.0 milestone (or, really, at every time needed) we merge 8.0 into 8.0-JDK17.

File: appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/node/PersistenceUnitNode.java [hunk 4]
```diff
@@ -113,6 +113,8 @@ private void initDispatchTable() {
         table.put(PersistenceTagNames.CLASS, "addClass");
         table.put(PersistenceTagNames.SHARED_CACHE_MODE, "setSharedCacheMode");
         table.put(PersistenceTagNames.VALIDATION_MODE, "setValidationMode");
+        table.put(PersistenceTagNames.SCOPE, "setValidationMode");
```
Comment (pzygielo) [inline appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/node/PersistenceUnitNode.java:116]: 👀 `setValidationMode` still?

File: appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/xml/PersistenceTagNames.java [hunk 5]
```diff
@@ -40,5 +40,7 @@ public interface PersistenceTagNames extends TagNames {
     public static String TRANSACTION_TYPE = "transaction-type";
     public static String SHARED_CACHE_MODE = "shared-cache-mode";
     public static String VALIDATION_MODE = "validation-mode";
+    public static String SCOPE = "scope";
+    public static String QAULIFIER = "qualifier";
```
Comment (pzygielo) [inline appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/xml/PersistenceTagNames.java:44]: :eyes: q**ua** or q**au**?

Comment (arjantijms) [reply appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/xml/PersistenceTagNames.java:44]: My guess would be "qua" being the right one, I'll change it

**Comment by arjantijms** [comment]
@starksm64 
>  Exception during lifecycle processing
java.lang.NullPointerException: Cannot invoke "java.lang.ClassLoader.loadClass(String)" because the return value of "com.sun.enterprise.deployment.BundleDescriptor.getClassLoader()" is null

Maybe a silly sounding suggestion, but did you simply try to compile GF with the changes once more? I build your PR, and I don't get that NPE at all. It just works.

**Comment by arjantijms** [comment]
@starksm64 

> In a CDI extension, obtain the entity manager configuration previously created, and add

At least this part works. I added a class to your test as follows:

```java
public class CdiExtension implements Extension  {

    public void register(@Observes BeforeBeanDiscovery beforeBean, BeanManager beanManager) {
        InvocationManager invocationManager = Globals.get(InvocationManager.class);

        Object container = invocationManager.getCurrentInvocation().getContainer();

        if (container instanceof ApplicationInfo applicationInfo) {

            PersistenceUnitsDescriptor persistenceUnitsDescriptor = applicationInfo.getMetaData(PersistenceUnitsDescriptor.class);

            for (PersistenceUnitDescriptor descriptor : persistenceUnitsDescriptor.getPersistenceUnitDescriptors()) {
                System.out.println("Description " + descriptor.getDescription());
                System.out.println("Qualifiers " + descriptor.getQualifiers());
                System.out.println("Scope " + descriptor.getScope());
            }
        }

    }
}
```

And the following dependencies (just for testing, this would go into GF itself):

```xml
        <dependency>
            <groupId>org.glassfish.main.common</groupId>
            <artifactId>glassfish-api</artifactId>
            <version>8.0.0-SNAPSHOT</version>
        </dependency>

        <dependency>
            <groupId>org.glassfish.main.common</groupId>
            <artifactId>internal-api</artifactId>
            <version>8.0.0-SNAPSHOT</version>
        </dependency>

        <dependency>
            <groupId>org.glassfish.main.deployment</groupId>
            <artifactId>dol</artifactId>
            <version>8.0.0-SNAPSHOT</version>
        </dependency>
```



The output is consistent with the data in `persistence.xml` of the test.

**Comment by starksm64** [comment]
> @starksm64
> 
> > Exception during lifecycle processing
> > java.lang.NullPointerException: Cannot invoke "java.lang.ClassLoader.loadClass(String)" because the return value of "com.sun.enterprise.deployment.BundleDescriptor.getClassLoader()" is null
> 
> Maybe a silly sounding suggestion, but did you simply try to compile GF with the changes once more? I build your PR, and I don't get that NPE at all. It just works.

What I just tried was:
1. Rebuild my fork
2. cp appserver/deployment/dol/target/dol.jar /Users/starksm/Dev/Jakarta/platform-tck/glassfish-runner/jpa-platform-tck/target/glassfish8/glassfish/modules/dol.jar
cp appserver/distributions/web/target/dependency/weld-integration.jar /Users/starksm/Dev/Jakarta/platform-tck/glassfish-runner/jpa-platform-tck/target/glassfish8/glassfish/modules/weld-integration.jar
cp appserver/web/web-glue/target/web-glue.jar /Users/starksm/Dev/Jakarta/platform-tck/glassfish-runner/jpa-platform-tck/target/glassfish8/glassfish/modules/web-glue.jar
3. run `mvn test-compile failsafe:integration-test@jpa-tests-cdi` in the platform-tck/glassfish-runner/jpa-platform-tck module


I'm still seeing the following NPE in the server.log:

```java
[2024-11-11T15:28:10.907404-06:00] [GF 8.0.0-JDK17-M7] [SEVERE] [NCLS-CORE-00026] [jakarta.enterprise.system.core] [tid: _ThreadID=53 _ThreadName=admin-listener(2)] [levelValue: 1000] [[
  Exception during lifecycle processing
java.lang.NullPointerException: Cannot invoke "java.lang.ClassLoader.loadClass(String)" because the return value of "com.sun.enterprise.deployment.BundleDescriptor.getClassLoader()" is null
	at com.sun.enterprise.deployment.archivist.Archivist.postStandardDDsRead(Archivist.java:367)
	at org.glassfish.web.deployment.archivist.WebArchivist.postStandardDDsRead(WebArchivist.java:179)
	at org.glassfish.web.deployment.archivist.WebArchivist.postStandardDDsRead(WebArchivist.java:72)
	at com.sun.enterprise.deployment.archivist.Archivist.readRestDeploymentDescriptors(Archivist.java:451)
	at com.sun.enterprise.deployment.archivist.Archivist.readDeploymentDescriptors(Archivist.java:426)
	at com.sun.enterprise.deployment.archivist.Archivist.open(Archivist.java:279)
	at com.sun.enterprise.deployment.archivist.Archivist.open(Archivist.java:288)
	at com.sun.enterprise.deployment.archivist.Archivist.open(Archivist.java:247)
	at com.sun.enterprise.deployment.archivist.ApplicationFactory.openArchive(ApplicationFactory.java:131)
	at org.glassfish.javaee.core.deployment.DolProvider.processDOL(DolProvider.java:187)
	at org.glassfish.javaee.core.deployment.DolProvider.load(DolProvider.java:210)
	at org.glassfish.javaee.core.deployment.DolProvider.load(DolProvider.java:80)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.loadDeployer(ApplicationLifecycle.java:942)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.setupContainerInfos(ApplicationLifecycle.java:881)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:411)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:257)
	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:471)
...
```

**Comment by arjantijms** [comment]
@starksm64 Another issue is maybe that we should start with a MUCH simpler test. The current test also doesn't seem to pass with just  

```java
@PersistenceUnit(unitName = "CTS-EM2")
private EntityManager em1;
```

On an unmodified GlassFish M7. Or am I missing something?

**Comment by arjantijms** [comment]
@starksm64 I found the issue, and overlooked it first. The test is wrong.

```java
@PersistenceUnit(unitName = "CTS-EM2")
private EntityManager em1;
```

Should be

```java
@PersistenceContext(unitName = "CTS-EM2")
private EntityManager em1;
```

`PersistenceUnit` is for injecting the factory.

**Comment by starksm64** [comment]
That was an initial commit. The test was updated to validate that using the
***@***.***(unitName = "CTS-EM")` does pass. Are you looking at
the issue1405 branch?

```bash
***@***.*** jpa-platform-tck % git remote -v
origin  https://github.com/jakartaee/platform-tck.git (fetch)
origin  https://github.com/jakartaee/platform-tck.git (push)
***@***.*** jpa-platform-tck % git status
On branch issue1405
Your branch is up to date with 'origin/issue1405'.
```

I'll look at trying to run against the full local snapshot build later
today.

**Comment by arjantijms** [comment]
> Are you looking at the issue1405 branch?

Indeed, that's the branch I'm looking at.

**Comment by starksm64** [comment]
The test code has been updated back to only using the custom annotations with inject.

I have rebuilt the glassfish fork and reran the test against it. I cannot get past the NPE issue.

```
mvn -Dmaven.build.cache.enabled=false -Pstaging clean install -Pfastest -T4C
...
mvn clean install -Dglassfish.container.version=8.0.0-SNAPSHOT -Dit.test=ServletEMLookupTest
...
Nov 12, 2024 11:48:56 PM org.omnifaces.arquillian.container.glassfish.clientutils.GlassFishClientUtil getResponseMap
SEVERE: While Deploying Application: 5c1268c2-6040-4a7b-ba29-0d117fe8db1f --exit_code: FAILURE, message: Error occurred during deployment: Exception while deploying the app [5c1268c2-6040-4a7b-ba29-0d117fe8db1f] : Cannot invoke "java.lang.ClassLoader.loadClass(String)" because the return value of "com.sun.enterprise.deployment.BundleDescriptor.getClassLoader()" is null. Please see server.log for more details. [status: CLIENT_ERROR reason: Bad Request]
Nov 12, 2024 11:48:56 PM org.omnifaces.arquillian.container.glassfish.clientutils.GlassFishClientUtil getResponseMap
WARNING: While Deploying Application: 5c1268c2-6040-4a7b-ba29-0d117fe8db1f --exit_code: FAILURE, message: Resource not found. [status: CLIENT_ERROR reason: Not Found]
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 7.626 s <<< FAILURE! -- in ee.jakarta.tck.persistence.ee.cdi.ServletEMLookupTest
[ERROR] ee.jakarta.tck.persistence.ee.cdi.ServletEMLookupTest -- Time elapsed: 7.626 s <<< ERROR!
org.jboss.arquillian.container.spi.client.container.DeploymentException: Could not deploy 5c1268c2-6040-4a7b-ba29-0d117fe8db1f.war
        at org.omnifaces.arquillian.container.glassfish.CommonGlassFishManager.deploy(CommonGlassFishManager.java:159)
        at org.omnifaces.arquillian.container.glassfish.managed.GlassFishManagedDeployableContainer.deploy(GlassFishManagedDeployableContainer.java:137)
        at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$3.call(ContainerDeployController.java:151)
        at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$3.call(ContainerDeployController.java:118)
        at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.executeOperation(ContainerDeployController.java:239)
        at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.deploy(ContainerDeployController.java:118)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
        at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
        at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
        at org.jboss.arquillian.container.impl.client.container.DeploymentExceptionHandler.verifyExpectedExceptionDuringDeploy(DeploymentExceptionHandler.java:47)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
        at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
        at org.jboss.arquillian.container.impl.client.ContainerDeploymentContextHandler.createDeploymentContext(ContainerDeploymentContextHandler.java:71)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
        at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
        at org.jboss.arquillian.container.impl.client.ContainerDeploymentContextHandler.createContainerContext(ContainerDeploymentContextHandler.java:54)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
        at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
        at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:134)
        at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:106)
        at org.jboss.arquillian.core.impl.EventImpl.fire(EventImpl.java:62)
        at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$1.perform(ContainerDeployController.java:92)
        at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$1.perform(ContainerDeployController.java:77)
        at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.forEachDeployment(ContainerDeployController.java:232)
        at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.forEachManagedDeployment(ContainerDeployController.java:212)
        at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.deployManaged(ContainerDeployController.java:77)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
        at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
        at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
        at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:134)
        at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:106)
        at org.jboss.arquillian.core.impl.EventImpl.fire(EventImpl.java:62)
        at org.jboss.arquillian.container.test.impl.client.ContainerEventController.execute(ContainerEventController.java:96)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
        at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
        at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
        at org.jboss.arquillian.test.impl.TestContextHandler.createClassContext(TestContextHandler.java:83)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
        at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
        at org.jboss.arquillian.test.impl.TestContextHandler.createSuiteContext(TestContextHandler.java:69)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
        at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
        at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:134)
        at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:106)
        at org.jboss.arquillian.test.impl.EventTestRunnerAdaptor.beforeClass(EventTestRunnerAdaptor.java:89)
        at org.jboss.arquillian.junit5.ArquillianExtension.beforeAll(ArquillianExtension.java:35)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: org.omnifaces.arquillian.jersey.server.ContainerException: While Deploying Application: 5c1268c2-6040-4a7b-ba29-0d117fe8db1f --exit_code: FAILURE, message: Error occurred during deployment: Exception while deploying the app [5c1268c2-6040-4a7b-ba29-0d117fe8db1f] : Cannot invoke "java.lang.ClassLoader.loadClass(String)" because the return value of "com.sun.enterprise.deployment.BundleDescriptor.getClassLoader()" is null. Please see server.log for more details. [status: CLIENT_ERROR reason: Bad Request]
        at org.omnifaces.arquillian.container.glassfish.clientutils.GlassFishClientUtil.getResponseMap(GlassFishClientUtil.java:312)
        at org.omnifaces.arquillian.container.glassfish.clientutils.GlassFishClientUtil.POSTMultiPartRequest(GlassFishClientUtil.java:215)
        at org.omnifaces.arquillian.container.glassfish.clientutils.GlassFishClientService.doDeploy(GlassFishClientService.java:324)
        at org.omnifaces.arquillian.container.glassfish.CommonGlassFishManager.deploy(CommonGlassFishManager.java:153)
        ... 54 more
```

**Comment by arjantijms** [comment]
> I have rebuilt the glassfish fork and reran the test against it. I cannot get past the NPE issue.

@starksm64 I've got everything working now using a CDI extension, in which I use the context class loader. See latest commit.

**Comment by arjantijms** [comment]
Things pass now, let's merge. I'll cross-merge to the 8.0 branch then.

**Pull request merged**